File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    private static int[] dx = {-1, 1, 0, 0};
    private static int[] dy = {0, 0, -1, 1};
    private static boolean[][] visited;
    private static Queue<int[]> queue = new LinkedList<>();
    
    private static int bfs(int[][] maps) {
        queue.add(new int[] {0, 0, 1});
        visited[0][0] = true;
        
        while(!queue.isEmpty()) {
            int[] poll = queue.poll();
            int x = poll[0];
            int y = poll[1];
            int z = poll[2];
            
            if(x == maps.length-1 && y == maps[0].length-1) {
                return z;
            }
            
            for(int d=0; d<4; d++) {
                int nx = x + dx[d];
                int ny = y + dy[d];
                
                if (nx < 0 || ny < 0 || nx >= maps.length || ny >= maps[0].length) continue;
                if (visited[nx][ny]) continue;
                
                if(maps[nx][ny] == 1) {
                    visited[nx][ny] = true;
                    queue.add(new int[] {nx, ny, z+1});
                }
            }
            
        }
        return -1;
    }
    
    public int solution(int[][] maps) {
        int answer = 0;
        visited = new boolean[maps.length][maps[0].length];
        answer = bfs(maps);
        
        return answer;
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    global answer
    answer = 0
    
    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]
    
    visited = [[False for _ in range(len(maps[0]))] for _ in range(len(maps))]
    
    def bfs():
        queue = deque([])
        queue.append([0, 0, 1])
        visited[0][0] = True
        while queue:
            x, y, c = queue.popleft()
            if x == len(maps) -1 and y == len(maps[0]) -1:
                return c
            for i in range(4):
                nx = x + dx[i]
                ny = y + dy[i]
                if nx < 0 or ny < 0 or nx >= len(maps) or ny >= len(maps[0]):
                    continue
                if not visited[nx][ny] and maps[nx][ny] == 1:
                    queue.append([nx, ny, c + 1])
                    visited[nx][ny] = True
        return -1
    
    answer = bfs()
    
    return answer